IoUring: enable completion injection

soulxu opened the pull request:
Commit Message: IoUring: enable completion injection
Additional Description:
This PR adds interfaces for add/remove injected completion to an iouring. At the same time, move the io_uring.h into the `envoy/`.

Risk Level: low
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

Signed-off-by: He Jie Xu [hejie.xu@intel.com](mailto:hejie.xu@intel.com)
Signed-off-by: Zhi Hao Xie [zhihao.xie@intel.com](mailto:zhihao.xie@intel.com)
Signed-off-by: Dmitry Rozhkov [dmitry.rozhkov@intel.com](mailto:dmitry.rozhkov@intel.com)

soulxu commented:
cc @KBaichoo @yanavlasov

KBaichoo commented on test/common/io/io_uring_impl_test.cc:
```diff
@@ -32,15 +32,15 @@ class IoUringImplTest : public ::testing::Test {
   testing::NiceMock<Server::Configuration::MockServerFactoryContext> context_;
   std::unique_ptr<IoUringFactoryImpl> factory_{};
+  OptRef<IoUring> io_uring_{};
   bool should_skip_{};
```
nit: const

KBaichoo reviewed (commented):
Generally looks good, thanks.

Assign @yanavlasov  for final pass

soulxu commented:
> Generally looks good, thanks.
> 
> Assign @yanavlasov for final pass

thanks for the review! updated as the comment, also addressed the comment from @yanavlasov in https://github.com/envoyproxy/envoy/pull/24958.

yanavlasov commented on envoy/common/io/io_uring.h:
```diff
@@ -12,8 +11,15 @@ namespace Io {
  * @param result is a return code of submitted system call.
+ * @param injected indicates whether the completion is injected or not.
+ */
+using CompletionCb = std::function<void(void* user_data, int32_t result, bool injected)>;
```
Right now C++ classes closely resemble the C io_uring API. Is it possible to make it more C++ and avoid the use of the `void*`. Conceptually io_uring allows submitting I/O operation and the completion callback. Would it make more sense for the `prepare*` calls to get the completion callback instead of the `void*` and then hide all of the `void*` business that C API needs inside the IoUring class?

soulxu replied:
That `user_data` is the data related to each request. I add C++ style `BaseRequest` to carry those user data. https://github.com/envoyproxy/envoy/pull/24958/files#diff-df9a08b7c36ffbd691c64a9c0c22a34cd2848af1ce434bc05bd661dbba4b7199R16

There are also a set of subclasses, like AcceptRequest, ReadRequest...etc 

The class `IoUring` is a pretty simple wrap of iouring C interface and low-level interface. so it still exposes that C-style stuff. I'm planning to hide all those details in the IoUringSocket and IoUringWorker interface. 
https://github.com/envoyproxy/envoy/pull/24958/files#diff-dad289fe5a60a8403503d6ec8e79baeb72f4de7a40aaba36646494cdfde1fd39R199
https://github.com/envoyproxy/envoy/pull/24958/files#diff-dad289fe5a60a8403503d6ec8e79baeb72f4de7a40aaba36646494cdfde1fd39R344

And I suppose only IoUringWorker will use the IoUring interface. All the up-layer code of Envoy only uses the IoUringWorker and IoUringSocket interface.

For the user of Iouring `IoUringSocketHandleImpl` won't see any `void*` and those C-style interfaces since it only uses the IoUringWorker and IoUringSocket.

Does that make sense?

soulxu replied:
Or I can change the interface like this 
`using CompletionCb = std::function<void(BaseRequest* user_data, int32_t result, bool injected)>;`

And prepare* also accept the `BaseRequest *`

yanavlasov replied:
Yes, the change you are proposing is good.

soulxu replied:
After thinking, let me change this in this PR, then I needn't the user data releasor anymore.

yanavlasov reviewed (commented):
/wait-any

soulxu commented on envoy/common/io/io_uring.h:
```diff
@@ -95,8 +101,31 @@ class IoUring {
+   * request completion by the upper-layer, then trigger the request completion processing.
+   * it is used to emulate an activation of READ/WRITE/CLOSED event on the specific file
+   * descriptor by the IoSocketHandle.
+   * @param fd is the file descriptor of this completion refer to.
```
Yea, this is used to implement https://github.com/envoyproxy/envoy/blob/59e0d299ee4715bde3d84171bfaeac5f478dd4b1/envoy/network/io_handle.h#L305

We need to activate the file event manually in many places, like
* When we have some data to send, we will trigger a write event https://github.com/envoyproxy/envoy/blob/59e0d299ee4715bde3d84171bfaeac5f478dd4b1/source/common/network/connection_impl.cc#L503
* When enabling the read, we may need manually trigger a read event to have the filter process the data in the buffer
https://github.com/envoyproxy/envoy/blob/59e0d299ee4715bde3d84171bfaeac5f478dd4b1/source/common/network/connection_impl.cc#L422

Let me update those to the commit message

yanavlasov replied:
OK, thanks makes sense.

yanavlasov commented on source/common/io/io_uring_impl.cc:
```diff
@@ -34,47 +34,76 @@ void IoUringFactoryImpl::onServerInitialized() {
     p.flags |= IORING_SETUP_SQPOLL;
   }
-  int ret = io_uring_queue_init_params(io_uring_size_, &ring_, &p);
+  // TODO (soulxu): According to the man page: `By default, the CQ ring will have twice the number
```
Is it possible that the queue can be starved and then prevent Envoy from making progress. For example, imagine the queue size of 10 and Envoy worker has 1 listening socket and 9 active TCP sockets. The queue may have 10 operations, 1 for listening socket and 9 for reading operations. Next Envoy wants to write into one of the sockets, it wants to place write operation but the queue is full. Will this cause problems?

soulxu replied:
For the SQ ring, we did see the ring can be full when there are too many iouring requests. Then the submission of new request will be rejected. Although we have configure option for the SQ ring size, that doesn't make sense for production use, since the user can't assume how much load the envoy will get. In the end, I'm thinking of some way to enable the envoy to wait until there is more room in the ring, then continue to submit the requests. I leave that as a todo https://github.com/envoyproxy/envoy/issues/28395, I want to improve that after we have iouring implementation first.

For the CQ ring, as the doc said, it will have twice the number of entries than the SQ ring default. At least for now, I don't have any idea for this now, I didn't see a case CQ ring is full in our tests. I also don't know in which case we will get twice the number of request completion that the request we submitted. But yes, we should figure that out, so I leave a todo here.

yanavlasov replied:
Ok, I understand. Reading a bit more on this, this just limits how many requests can be submitted in one go, not the overall limit of kernel queue. It is a bit confusing with multiple different queue size values in io_uring_queue_init and io_uring_setup. 

It looks to me that what this code should never overflow the max size of the submission batch, since we never process more than io_uring_size_ of operations one wake up. If some test fails, maybe the value passed to the io_uring_queue_init_params just needs some overhead added. Not sure what it needs to be though.

soulxu replied:
> It looks to me that what this code should never overflow the max size of the submission batch, since we never process more than io_uring_size_ of operations one wake up. If some test fails, maybe the value passed to the io_uring_queue_init_params just needs some overhead added. Not sure what it needs to be though.

As I remember some tests failed flaky due to the uring size. Maybe I should add some stat to see how many requests in one wake up, then we will know what happened. Thanks for the reminder, I will keep this in mind, and figure out in the future.

yanavlasov reviewed (commented):
/wait-any

soulxu commented:
@phlax do you know why the coverage test failed some files I didn't change?

yanavlasov commented:
Please merge main. Coverage was broken for a small range of commits and is fixed now.

soulxu commented:
> Please merge main. Coverage was broken for a small range of commits and is fixed now.

thanks, done

soulxu commented:
/retest
